autoprotocol/builders.py

```diff
@@ -1563,30 +1579,70 @@ def device_mode_params(
-            if model is None:
-                model = "high_volume"
-            if chip_material is None:
-                chip_material = "pfe"
-            if nozzle is None:
-                nozzle = "standard"
-            x_tempest_chip = {
-                "model": model,
-                "material": chip_material,
-                "nozzle": nozzle,
+        if device is "x_mantis":
+            default_params: dict = {
+                "model": "high_volume",
+                "material": None,
+                "nozzle": None,
+                "diaphragm": 0,
+                "nozzle_size": "0.1:mm",
+                "tubing": "LV",
+                "z_drop": "0.0:mm",
+                "viscosity": "1",
+            }
+            accepted_params: dict = {
+                "model": ["high_volume", "low_volume"],
+                "diaphragm": [0, 100],
+                "nozzle_size": ["0.1:mm", "0.2:mm", "0.5:mm"],
+                "tubing": ["LV", "HV", "P200", "P1000"],
+                "z_drop": ["0.0:mm", "100.0:mm"],
+                "viscosity": ["1", "2-5", "6-10", "11-20", "21-25"],
+            }
+        else:
+            default_params: dict = {
+                "model": "high_volume",
+                "material": "pfe",
+                "nozzle": "standard",
+                "diaphragm": None,
+                "nozzle_size": None,
+                "tubing": None,
+                "z_drop": None,
+                "viscosity": None,
             }
-            device_mode_params["x_tempest_chip"] = x_tempest_chip
+            accepted_params: dict = {
+                "model": ["high_volume"],
+                "nozzle": ["standard"],
+                "material": ["silicone", "pfe"],
+            }
+        device_mode_params = {}
+        device_dict: dict = {}
+        if any([model, material, nozzle, diaphragm, nozzle_size, tubing, z_drop, viscosity]):
+            device_dict.update({
+                "model": model or default_params["model"],
+                "material": material or default_params["material"],
+                "nozzle": nozzle or default_params["nozzle"],
+                "diaphragm": diaphragm or default_params["diaphragm"],
+                "nozzle_size": nozzle_size or default_params["nozzle_size"],
+                "tubing": tubing or default_params["tubing"],
+                "z_drop": z_drop or default_params["z_drop"],
+                "viscosity": viscosity or default_params["viscosity"],
+            })
+        error_values: dict = {}
+        # Validate params with accepted params dict
+        for key, value in device_dict.items():
+            if value and value not in accepted_params[key]:
+                error_values.update({key: value})
+
+        if error_values:
+            raise ValueError(
+                f"Incorrect params: {error_values}. It must be {accepted_params}"
+            )
+        else:
+            device_mode_params[device] = device_dict
```

Comment:
you can just return this since the raised error would handle incorrect params. One way to do this is removing the initialized dict and just returning {device: device_dict}

Reply:
ahh yeah ok!